### PR TITLE
Add block sync time estimate

### PR DESCRIPTION
**Pull request by grod220**
Part 1 of https://github.com/penumbra-zone/web/issues/305. Changes extension preview. Next PR will be the web app update.

Calculates time remaining on sync via _Exponential Weighted Moving Average_

https://github.com/penumbra-zone/web/assets/16624263/87d93357-3bf7-473a-8bf0-9f5ffb7af734

### CODE REVIEW

File: packages/ui/components/ui/block-sync-status/index.tsx
Comment (grod220) [inline packages/ui/components/ui/block-sync-status/index.tsx:1]: Largely copied over from the old code

File: packages/ui/components/ui/block-sync-status/sync-progress-hook.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      setConfident(true);
+    }
+  }, [lastBlockSynced, syncUpdates, syncUpdatesThreshold, confident]);
+
+  const blocksRemaining = lastBlockHeight - lastBlockSynced;
+  const timeRemaining = speed > 0 ? blocksRemaining / speed : Infinity;
+  const formattedTimeRemaining =
+    timeRemaining === Infinity ? '' : humanizeDuration(timeRemaining * 1000, { round: true });
+
+  return { speed, timeRemaining, formattedTimeRemaining, confident };
```
Comment (jessepinho) [inline packages/ui/components/ui/block-sync-status/sync-progress-hook.ts:55]:
```suggestion
  return { formattedTimeRemaining, confident };
```

(Right? Or will you be using those other values in the web app or something?)

File: packages/ui/components/ui/block-sync-status/sync-progress-hook.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import humanizeDuration from 'humanize-duration';
+
+/**
+ * Custom hook to calculate synchronization speed and estimate the time remaining
+ * for a synchronization process to complete, using the Exponential Weighted Moving Average (EWMA)
+ * to smooth the speed calculation.
+ *
+ * @returns An object containing:
+ *          - speed: Current speed of synchronization in blocks per second.
+ *          - timeRemaining: Estimated time remaining to complete the synchronization, in seconds.
```
Comment (jessepinho) [inline packages/ui/components/ui/block-sync-status/sync-progress-hook.ts:12]: See my comment below — are these values needed?

**Review by jessepinho — APPROVED** [review]
Very cool! LGTM, although some tests would be nice

File: packages/ui/components/ui/block-sync-status/ewma.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * However, the reason why we have our own code for this is because this library
+ * was missing types and can only run in nodejs, not a browser environment.
+ *
+ * Ewma (Exponential Weighted Moving Average) is used to calculate a smoothed average of a series
+ * of numbers, where more recent values have a greater weight than older values. This class is
+ * particularly useful for smoothing out time-series data or calculating an average that needs to
+ * adapt quickly to changes. The half-life parameter controls how quickly the weights decrease for
+ * older values.
+ */
+export class EWMA {
```
[the remark this answers is not in the document]
Comment (grod220) [reply packages/ui/components/ui/block-sync-status/ewma.ts:23]: Added 👍

**Pull request merged**